sql/handler.cc

```diff
@@ -2792,7 +2802,11 @@ static my_bool xarecover_handlerton(THD *unused, plugin_ref plugin,
         }
         else if (!info->mem_root)
         {
-          int rc= hton->rollback_by_xid(hton, info->list+i);
+          int rc= 0;
+          if (hton->recover_rollback_by_xid)
+            rc= hton->recover_rollback_by_xid(hton, info->list+i);
+          else
+            rc= hton->rollback_by_xid(hton, info->list+i);
```

Comment:
I think that we must assert or check here if `wsrep_is_wsrep_xid(info->list+i)` holds. If yes, then we’d better invoke `hton->rollback_by_xid()` in the old-fashioned way, to avoid any race condition.

Reply:
done